File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.java
import java.util.*;
import java.io.*;
public class Main {

    public static int change(char target) {
        if (target == 'A') {
            return 0;
        } else if (target == 'C') {
            return 1;
        } else if (target == 'G') {
            return 2;
        } else if (target == 'T') {
            return 3;
        }
        return -1;
    }

    public static void main(String [] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int dN = Integer.parseInt(st.nextToken());
        int sN = Integer.parseInt(st.nextToken());

        char [] dna = br.readLine().toCharArray();

        st = new StringTokenizer(br.readLine());
        int [] acgt= new int[4];
        for(int i=0; i<4; i++){
            acgt[i] = Integer.parseInt(st.nextToken());
        }

        int start = 0;
        int end = start + sN -1;
        int [] count = new int[4];

        for(int i=start; i<=end; i++){
            int index = change(dna[i]);
            if(index!=-1){
                count[index]+=1;
            }
        }

        int answer=0;
        while (end < dN){
            int check=0;
            for(int i=0; i<4; i++){
                if(acgt[i] <= count[i]){
                    check+=1;
                }
            }
            if(check==4) {
                answer += 1;
            }

            if(end==dN-1){
                break;
            }

            int s_index = change(dna[start]);
            if (s_index!=-1){
                count[s_index]-=1;
            }
            start+=1;
            end+=1;
            int e_index = change(dna[end]);
            if (e_index!=-1) {
                count[e_index] += 1;
            }

        }
        System.out.println(answer);
    }
}

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
# DNA 비밀번호(실버2)
import sys
input = sys.stdin.readline
def check(needs, state):
    answer=0
    for i in range(4):
        if needs[i] <= state[i]:
            answer+=1
    if answer == 4:
        return True
    else:
        return False
    
dna_len, str_len = map(int, input().split())
dna = list(input())
needs = list(map(int, input().split()))
start=0
end = str_len-1
state = [0]*(4)
total=0
for i in range(start, end+1):
    if dna[i] == 'A':
        state[0]+=1
    elif dna[i] == 'C':
        state[1]+=1
    elif dna[i] == 'G':
        state[2]+=1
    else:
        state[3]+=1

if check(needs, state):
    total+=1

while end < dna_len-1:
    
    if dna[start]=='A':
        state[0]-=1
    elif dna[start]=='C':
        state[1]-=1
    elif dna[start]=='G':
        state[2]-=1
    else:
        state[3]-=1
    
    start+=1
    end+=1
    if dna[end] == 'A':
        state[0]+=1
    elif dna[end] == 'C':
        state[1]+=1
    elif dna[end] == 'G':
        state[2]+=1
    else:
        state[3]+=1
    
    if check(needs, state):
        total+=1

print(total)